solr/core/src/test/org/apache/solr/store/shared/SharedCoreIndexingBatchProcessorTest.java

```diff
@@ -0,0 +1,270 @@
+   */
+  @Test
+  public void testUpToDateCoreIsNotPulledAtStart() throws Exception {
+    SharedShardMetadataController.SharedShardVersionMetadata shardVersionMetadata = core.getCoreContainer()
+        .getSharedStoreManager().getSharedShardMetadataController().readMetadataValue(COLLECTION_NAME, SHARD_NAME);
+    ClusterState clusterState = core.getCoreContainer().getZkController().getClusterState();
+    DocCollection collection = clusterState.getCollection(COLLECTION_NAME);
+    String sharedShardName = (String) collection.getSlicesMap().get(SHARD_NAME).get(ZkStateReader.SHARED_SHARD_NAME);
+    corePuller.pullCoreFromSharedStore(core, sharedShardName, shardVersionMetadata, true);
+    verify(corePuller).pullCoreFromSharedStore(any(), any(), any(), anyBoolean());
+    verify(processor, never()).startIndexingBatch();
+    processAddOrDelete();
+    verify(processor).startIndexingBatch();
+    verify(corePuller).pullCoreFromSharedStore(any(), any(), any(), anyBoolean());
+  }
+
+  /**
+   * Tests that a read lock is acquired even when the start encounters an error.
+   */
+  @Test
+  public void testReadLockIsAcquiredEvenStartEncountersError() throws Exception {
+    doThrow(new SolrException(SolrException.ErrorCode.SERVER_ERROR, "pull failed"))
+        .when(corePuller).pullCoreFromSharedStore(any(), any(), any(), anyBoolean());
+    verify(processor, never()).startIndexingBatch();
+    verify(corePuller, never()).pullCoreFromSharedStore(any(), any(), any(), anyBoolean());
+    assertEquals("wrong pull read lock count", 0, corePullLock.getReadLockCount());
+    boolean errorOccurred = false;
+    try {
+      processor.startIndexingBatch();
+      fail("No exception thrown");
+    } catch (Exception ex) {
+      assertTrue("Wrong exception thrown", ex.getMessage().contains("pull failed"));
+    }
+    assertEquals("wrong pull read lock count", 1, corePullLock.getReadLockCount());
+    verify(processor).startIndexingBatch();
+    verify(corePuller).pullCoreFromSharedStore(any(), any(), any(), anyBoolean());
+  }
+
+  /**
+   * Tests that an indexing batch with some work does push to the shared store.
+   */
+  @Test
+  public void testCommitAfterAddOrDeleteDoesPush() throws Exception {
+    processAddOrDelete();
+    processCommit();
+    verify(corePusher).pushCoreToSharedStore(any(), any());
+  }
+
+  /**
+   * Tests that an indexing batch with no work does not push to the shared store.
+   */
+  @Test
+  public void testIsolatedCommitDoesNotPush() throws Exception {
+    processCommit();
+    verify(corePusher, never()).pushCoreToSharedStore(any(), any());
+  }
+
+  /**
+   * Tests that an already committed indexing batch throws if a doc/deleted again.
+   */
+  @Test
+  public void testAddOrDeleteAfterCommitThrows() throws Exception {
+    processAddOrDelete();
+    testAddOrDeleteAfterIsolatedCommitThrows();
+  }
+
+  /**
+   * Tests that an already isolated committed indexing batch throws if a doc/deleted again.
+   */
+  @Test
```

Comment:
is this supposed to be its own test from L219?

Reply:
Two different flavors of tests. One with commit after some add/delete and the other one just isolated commit. I also got confused re-reading this. I believe confusion was coming from the re-use of test method. To make it more readable I have extracted the common code in separate method.